### ISSUE TITLE
[Idea]: Functions for numerical integration and differentiation

### ISSUE THREAD

**Issue by Planeshifter** [stdlib-js/google-summer-of-code#24]
### Idea

The goal of this idea is to add functions for numerical integration or differentiation to stdlib-js as building blocks for downstream algorithms. The functions could be ported from permissively licensed open-source libraries in other languages such as C or Fortran or alternatively be implemented from scratch by consulting the literature and reference implementations from various languages.

### Expected Outcomes

stdlib will have a range of robust functions for performing numerical integration or differentiation

### Involved Software

No other software is necessary.

### Prerequisite Knowledge

JavaScript, Node.js.

### Difficulty

Intermediate.

### Project Length

350 hours.

**Comment by rreusser** [comment]
After writing a lot of, honestly, pretty bad code in this domain, my general feeling is that the most difficult aspect of this task may just be the public-facing API design. There are lots of interesting extensions like [poincaré maps](https://en.wikipedia.org/wiki/Poincar%C3%A9_map), for example, or stopping criteria which aren't difficult to design, but writing a coherent API for various methods and their extensions does require some careful thought.

**Comment by kgryte** [comment]
@rreusser Agreed. Were a GSoC contributor interested in this idea, we'd need to spend some time up front sketching out the API design and probably doing a few POC packages.
